File: app/src/main/java/ethiopia/covid/android/ui/fragment/NewsFragment.java
package ethiopia.covid.android.ui.fragment;

import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.stfalcon.imageviewer.StfalconImageViewer;

import java.util.ArrayList;
import java.util.Collections;

import ethiopia.covid.android.App;
import ethiopia.covid.android.R;
import ethiopia.covid.android.data.NewsItem;
import ethiopia.covid.android.ui.activity.MainActivity;
import ethiopia.covid.android.ui.adapter.NewsItemRecyclerAdapter;
import ethiopia.covid.android.util.Utils;
import me.ibrahimsn.lib.Badge;
import me.ibrahimsn.lib.BadgeType;
import timber.log.Timber;

import static ethiopia.covid.android.ui.fragment.ContentState.changeErrorDialogVisibility;
import static ethiopia.covid.android.ui.fragment.ContentState.changeProgressBarVisibility;
import static ethiopia.covid.android.ui.fragment.ContentState.setRefreshButtonAction;
import static ethiopia.covid.android.util.Constant.PREFERENCE_LATEST_NEWS;
import static java.lang.Math.min;

/**
 * Created by BrookMG on 3/23/2020 in ethiopia.covid.android.ui.fragment
 * inside the project CoVidEt .
 */
public class NewsFragment extends BaseFragment {

    private RecyclerView mainNewsRecycler;
    private NewsItemRecyclerAdapter adapter;

    public static NewsFragment newInstance() {
        Bundle args = new Bundle();
        NewsFragment fragment = new NewsFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @RequiresApi(api = Build.VERSION_CODES.KITKAT_WATCH)
    private void handleWindowInsets(RecyclerView recyclerView) {
        recyclerView.setOnApplyWindowInsetsListener((v1, insets) -> {
            ViewGroup.MarginLayoutParams marginParams = (ViewGroup.MarginLayoutParams) recyclerView.getLayoutParams();
            marginParams.setMargins( 0, insets.getSystemWindowInsetTop(), 0, 0);
            recyclerView.setLayoutParams(marginParams);

            return insets;
        });
    }

    private void onLoadMoreNews() {
        int currentLastNewsId = adapter.idOfLastNews();
        App.getInstance().getMainAPI().getLatestNews(currentLastNewsId , (item, err) -> {
            if (!err.isEmpty()) Timber.e(err);
            adapter.loadMoreOnBottom(item);
        });
    }

    @Override
    public void onStart() {
        super.onStart();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT_WATCH) {
            mainNewsRecycler.requestApplyInsets();
            handleWindowInsets(mainNewsRecycler);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mainNewsRecycler = null;
        adapter = null;
    }

    @Override
    public void onReselect() {
        super.onReselect();
        // Go to the top of the recycler view
        mainNewsRecycler.smoothScrollToPosition(0);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View mainView = inflater.inflate(R.layout.news_fragment, container, false);
        mainNewsRecycler = mainView.findViewById(R.id.recycler_view);

        mainNewsRecycler.setLayoutManager(new LinearLayoutManager(getActivity() , RecyclerView.VERTICAL , false));

        setRefreshButtonAction(mainView , v -> renderNews(mainView));
        renderNews(mainView);

        return mainView;
    }

    private void renderNews(View mainView) {
        adapter = new NewsItemRecyclerAdapter(new ArrayList<>(), this::onLoadMoreNews , (mainImageView, clickedImageUri) -> {
            if (getActivity() != null)
            new StfalconImageViewer.Builder<>(getActivity(), clickedImageUri,
                    (imageView, image) -> Glide.with(imageView).load(image).into(imageView))
                    .withTransitionFrom(mainImageView)
                    .show(true);
        });

        changeProgressBarVisibility(mainView , true);
        changeErrorDialogVisibility(mainView, false);

        App.getInstance().getMainAPI().getLatestNews((item, err) -> {
            int previousLatestNews = PreferenceManager.getDefaultSharedPreferences(App.getInstance())
                    .getInt(PREFERENCE_LATEST_NEWS , 0);

            if (item == null) {
                changeProgressBarVisibility(mainView , false);
                changeErrorDialogVisibility(mainView, true);
                return;
            }

            // If the news is not found in the current incoming list ...
            // That means the user missed more than 10 news
            int currentNewsPosition = 11;

            for (int i = 0; i < item.size(); i++) {
                if (item.get(i).getId() == previousLatestNews) {
                    currentNewsPosition = i;
                    break;
                }
            }

            showBadge(currentNewsPosition);

            PreferenceManager.getDefaultSharedPreferences(App.getInstance())
                    .edit().putInt(PREFERENCE_LATEST_NEWS , item.get(0).getId())
                    .apply();

            if (adapter != null) adapter.populateNewsItems(item);
            changeProgressBarVisibility(mainView , false);
        });

        mainNewsRecycler.setAdapter(adapter);
    }

    private void showBadge(int numberOfNewNews) {
        if (numberOfNewNews <= 0) return;

        if (getActivity() instanceof MainActivity) {
            ((MainActivity) getActivity()).setBadge(2 , new Badge(
                    20F,
                    min(numberOfNewNews, 10) + (numberOfNewNews > 10 ? "+" : ""),
                    ContextCompat.getColor(getActivity() , R.color.yellow_0),
                    Color.BLACK,
                    8F,
                    BadgeType.BOX
            ));
        }

        new Handler().postDelayed(() -> {
            if (getActivity() instanceof MainActivity) ((MainActivity) getActivity()).removeBadge(2);
        } , 10_000);
    }


}

File: app/src/main/java/ethiopia/covid/android/ui/fragment/NewsFragment.kt
package ethiopia.covid.android.ui.fragment

import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.MarginLayoutParams
import android.view.WindowInsets
import android.widget.ImageView
import androidx.annotation.RequiresApi
import androidx.core.content.ContextCompat
import androidx.preference.PreferenceManager.getDefaultSharedPreferences
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.stfalcon.imageviewer.StfalconImageViewer
import com.stfalcon.imageviewer.loader.ImageLoader
import ethiopia.covid.android.App.Companion.instance
import ethiopia.covid.android.R
import ethiopia.covid.android.data.NewsItem
import ethiopia.covid.android.databinding.ContentStateLayoutBinding
import ethiopia.covid.android.databinding.NewsFragmentBinding
import ethiopia.covid.android.network.API.OnItemReady
import ethiopia.covid.android.ui.activity.MainActivity
import ethiopia.covid.android.ui.adapter.NewsItemRecyclerAdapter
import ethiopia.covid.android.ui.adapter.NewsItemRecyclerAdapter.OnLastItemReachedListener
import ethiopia.covid.android.ui.adapter.OnImageItemClicked
import ethiopia.covid.android.ui.fragment.ContentState.changeErrorDialogVisibility
import ethiopia.covid.android.ui.fragment.ContentState.changeProgressBarVisibility
import ethiopia.covid.android.ui.fragment.ContentState.setRefreshButtonAction
import ethiopia.covid.android.util.Constant.PREFERENCE_LATEST_NEWS
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import me.ibrahimsn.lib.Badge
import me.ibrahimsn.lib.BadgeType
import timber.log.Timber
import java.util.*

/**
 * Created by BrookMG on 3/23/2020 in ethiopia.covid.android.ui.fragment
 * inside the project CoVidEt .
 */
class NewsFragment : BaseFragment() {
    private var _newsFragmentBinding: NewsFragmentBinding? = null
    private val newsFragmentBinding: NewsFragmentBinding
        get() = _newsFragmentBinding!!

    private var adapter: NewsItemRecyclerAdapter? = null

    @RequiresApi(api = Build.VERSION_CODES.KITKAT_WATCH)
    private fun handleWindowInsets(recyclerView: RecyclerView?) {
        recyclerView?.setOnApplyWindowInsetsListener { _: View?, insets: WindowInsets ->
            val marginParams = recyclerView.layoutParams as MarginLayoutParams
            marginParams.setMargins(0, insets.systemWindowInsetTop, 0, 0)
            recyclerView.layoutParams = marginParams
            insets
        }
    }

    private fun onLoadMoreNews() {
        val currentLastNewsId = adapter?.idOfLastNews()

        if (currentLastNewsId != null)
        instance.mainAPI.getLatestNews(currentLastNewsId, object : OnItemReady<List<NewsItem>?> {
            override fun onItem(item: List<NewsItem>?, err: String?) {
                if (err?.isNotEmpty() == true) Timber.e(err)
                adapter?.loadMoreOnBottom(item)
            }
        })
    }

    override fun onStart() {
        super.onStart()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT_WATCH) {
            newsFragmentBinding.recyclerView.requestApplyInsets()
            handleWindowInsets(newsFragmentBinding.recyclerView)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _newsFragmentBinding = null
        adapter = null
    }

    override fun onReselect() {
        super.onReselect()
        // Go to the top of the recycler view
        newsFragmentBinding.recyclerView.smoothScrollToPosition(0)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        _newsFragmentBinding = NewsFragmentBinding.inflate(layoutInflater)

        newsFragmentBinding.recyclerView.layoutManager = LinearLayoutManager(activity, RecyclerView.VERTICAL, false)
        setRefreshButtonAction(newsFragmentBinding.contentState, View.OnClickListener { renderNews(newsFragmentBinding.contentState) })
        renderNews(newsFragmentBinding.contentState)
        return newsFragmentBinding.root
    }



    private fun renderNews(mainView: ContentStateLayoutBinding) {
        adapter = NewsItemRecyclerAdapter(ArrayList(), object : OnLastItemReachedListener {
            override fun onLastItemReached() {
                onLoadMoreNews()
            }
        }, object : OnImageItemClicked {
            override fun onImageClicked(imageView: ImageView, clickedImageUri: List<String>) {
                if (activity != null)
                    StfalconImageViewer.Builder(activity, clickedImageUri,
                        ImageLoader { imageViewX: ImageView?, image: String? ->
                            run {
                                if (imageViewX != null)
                                    Glide.with(imageViewX).load(image).into(imageViewX)
                            }
                        })
                        .withTransitionFrom(imageView)
                        .show(true)
            }
        })

        changeProgressBarVisibility(mainView, true)
        changeErrorDialogVisibility(mainView, false)

        instance.mainAPI.getLatestNews(object : OnItemReady<List<NewsItem>?> {
            override fun onItem(item: List<NewsItem>?, err: String?) {
                if (item == null || item.isEmpty()) {
                    changeProgressBarVisibility(mainView, false)
                    changeErrorDialogVisibility(mainView, true)
                    return
                }

                CoroutineScope(Dispatchers.Default).launch {
                    val previousLatestNews = getDefaultSharedPreferences(instance)
                            .getInt(PREFERENCE_LATEST_NEWS, 0)

                    // If the news is not found in the current incoming list ...
                    // That means the user missed more than 10 news
                    var currentNewsPosition = 11
                    for (i in item.indices) {
                        if (item[i].id == previousLatestNews) {
                            currentNewsPosition = i
                            break
                        }
                    }

                    getDefaultSharedPreferences(instance)
                            .edit().putInt(PREFERENCE_LATEST_NEWS, item[0].id)
                            .apply()

                    withContext(Dispatchers.Main) {
                        showBadge(currentNewsPosition)
                        if (adapter != null) adapter?.populateNewsItems(item)
                        changeProgressBarVisibility(mainView, false)
                    }
                }
            }
        })

        newsFragmentBinding.recyclerView.adapter = adapter
    }

    private fun showBadge(numberOfNewNews: Int) {
        if (numberOfNewNews <= 0) return

        (activity as? MainActivity)?.setBadge(2, Badge(
                20f,
                numberOfNewNews.coerceAtMost(10).toString() + if (numberOfNewNews > 10) "+" else "",
                ContextCompat.getColor(activity as MainActivity, R.color.yellow_0),
                Color.BLACK,
                8f, BadgeType.BOX))

        Handler().postDelayed({ (activity as? MainActivity)?.removeBadge(2) }, 10000)
    }

    companion object {
        fun newInstance(): NewsFragment {
            val args = Bundle()
            val fragment = NewsFragment()
            fragment.arguments = args
            return fragment
        }
    }
}